tests/test_ell1h.py

```diff
@@ -12,111 +13,223 @@
-                assert np.nanmax(relative_diff) < tol, msg
+# TODO need a better derivative test
+def test_derivative(toasJ1853, modelJ1853):
+    log = logging.getLogger("TestJ1853.derivative_test")
+    test_params = ["H3", "H4", "STIGMA"]
+    modelJ1853.H4.value = 0.0  # For test PBDOT
+    modelJ1853.STIGMA.value = 0.0
+    testp = tdu.get_derivative_params(modelJ1853)
+    delay = modelJ1853.delay(toasJ1853)
+    # Change parameter test step
+    testp["H3"] = 6e-1
+    testp["H4"] = 1e-2
+    testp["STIGMA"] = 1e-2
+    for p in test_params:
+        log.debug("Runing derivative for %s", "d_delay_d_" + p)
+        ndf = modelJ1853.d_phase_d_param_num(toasJ1853, p, testp[p])
+        adf = modelJ1853.d_phase_d_param(toasJ1853, delay, p)
+        diff = adf - ndf
+        if not np.all(diff.value) == 0.0:
+            mean_der = (adf + ndf) / 2.0
+            relative_diff = np.abs(diff) / np.abs(mean_der)
+            # print "Diff Max is :", np.abs(diff).max()
+            msg = (
+                "Derivative test failed at d_delay_d_%s with max relative difference %lf"
+                % (p, np.nanmax(relative_diff).value)
+            )
+            if p in ["EPS1DOT", "EPS1"]:
+                tol = 0.05
             else:
-                continue
-
-    def test_J0613_H4(self):
-        log = logging.getLogger("TestJ0613.fit_tests")
-        f = ff.GLSFitter(self.toasJ0613, self.modelJ0613)
-        f.fit_toas()
-        f.model.free_params = ("H3", "H4")
-        for pn, p in f.model.get_params_dict("free", "quantity").items():
-            op = getattr(f.model_init, pn)
-            diff = np.abs(p.value - op.value)
-            sigma = diff / op.uncertainty_value
-            # Fit th
-            assert (
-                sigma < 0.7
-            ), "refit %s is %lf sigma different from original value" % (pn, sigma)
-
-    def test_J0613_STIG(self):
-        log = logging.getLogger("TestJ0613.fit_tests_stig")
-        f = ff.GLSFitter(self.toasJ0613, self.modelJ0613_STIG)
-        f.fit_toas()
-        f.model.free_params = ("H3", "STIGMA")
-        for pn, p in f.get_params_dict("free", "quantity").items():
-            op = getattr(f.model_init, pn)
-            diff = np.abs(p.value - op.value)
-            sigma = diff / op.uncertainty_value
-            # Fit th
-            assert (
-                sigma < 0.7
-            ), "refit %s is %lf sigma different from original value" % (pn, sigma)
-
-
-if __name__ == "__main__":
-    pass
+                tol = 1e-3
+            log.debug(
+                "derivative relative diff for %s, %lf"
+                % ("d_delay_d_" + p, np.nanmax(relative_diff).value)
+            )
+            assert np.nanmax(relative_diff) < tol, msg
+        else:
+            continue
+
+
+def test_J0613_H4(toasJ0613, modelJ0613):
+    log = logging.getLogger("TestJ0613.fit_tests")
+    f = ff.GLSFitter(toasJ0613, modelJ0613)
+    f.fit_toas()
+    f.model.free_params = ("H3", "H4")
```

Comment:
This line does not appear to do anything since the `fit_toas` call is before this?

Reply:
Fitting H3 and H4 changes A1 and EPS1 3 to 4 sigma and caused a failure on the parameter change. The value of H3 and H4 are derived from sini and M2.

Reply:
Er, what I meant was: setting the free parameters *after* the fitting instead of before doesn't do anything and seems like a mistake?